### PR TITLE
fixes #21171; dynamic acyclic refs need to use dyn decRef

### LINKED ISSUE #21171: Fatal memory overwrite bug with `acyclic` pragmas
### Description

I am experiencing an issue where it seems that sometimes a reference (pointer) is overwritten to nil
It appears to occur (thank you @ringabout ) when using the acyclic pragma in ORC. Which makes me believe this is a bug.

It is relatively hard to reproduce. I tried to make the minimal version of my program that creates the issue.

In this sample code, I have a node hierarchy with templates to construct a node tree.
When this code is ran, after a few thousand succesful iterations, the root node has a 'nil' pointer in it's list of children.
This then causes the dynamic dispatch to error because it cannot determine the type of the child node.

This issue is super weird and again, only happens on ORC. I would really like if someone helped me understand what is going on, and whether or not this is a serious bug in nim.

The issue can be avoided if you don't use the acyclic pragma, don't manually allocate memory (which seems to be responsible for overwriting the pointer) or if you use a template for the textNode constructor instead of a proc.

Either way, if the textNode inserts itself into it's parents children list it should not be overwritten when calling alloc0, because it should still be referenced.

### Nim Version

Nim Compiler Version 1.9.1 [Linux: amd64]
Compiled at 2022-12-24
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: 86a6c90c9e23f2e79d9d43927e114eedec0c230d
active boot switches: -d:release

### Current Output

```text
Hint: /home/lennart/Projects/nim-game/reftest [Exec]
/home/lennart/Projects/nim-game/reftest.nim(71) reftest
/home/lennart/Projects/nim-game/reftest.nim(56) doTest
/home/lennart/Projects/nim-game/reftest.nim(32) printChildren
/home/lennart/.choosenim/toolchains/nim-#devel/lib/system/chcks.nim(105) printNode
/home/lennart/.choosenim/toolchains/nim-#devel/lib/system/fatal.nim(54) sysFatal
Error: unhandled exception: cannot dispatch; dispatcher is nil [NilAccessDefect]
Error: execution of an external program failed: '/home/lennart/Projects/nim-game/reftest'
```

### Expected Output

```text
Nothing
```

### Possible Solution

I believe this potential bug deserves further diagnosis before a solution can be proposed.

### Additional Information

I created the following program to reproduce the bug.
I am sorry for the lengthy code, but it is still substantially more simple than my own program which uses lots of dynamic libraries.

```nim
import std/[with]

type
  Node* {.acyclic.} = ref object of RootObj
    name: string
    data: pointer
    children: seq[Node]
  TextNode = ref object of Node
    text: string

proc fakeEcho(s: string) =
  if s.len < 0:
    echo s

proc newNode[T: Node](parent: Node): T =
  new result
  result.data = alloc0(250)
  parent.children.add(result)

proc newRootNode(): Node =
  new result
  result.data = alloc0(250)

method printNode(node: Node) {.base.} =
  fakeEcho node.name

method printNode(node: TextNode) =
  procCall printNode(Node(node))
  fakeEcho node.text

proc printChildren(node: Node) =
  for child in node.children:
    child.printNode()
    printChildren(child)

proc free(node: Node) =
  for child in node.children:
    free(child)
  dealloc(node.data)

template node(parent: Node, body: untyped): untyped =
  var node = newNode[Node](parent)
  with node:
    body

proc textNode(parent: Node, text: string) =
  var node = newNode[TextNode](parent)
  node.text = text

template withRootNode(body: untyped): untyped =
  var root = newRootNode()
  root.name = "root"
  with root:
    body
  root.printNode()
  printChildren(root)
  root.free()

proc doTest() =
  withRootNode:
    node:
      name = "child1"
      node:
        name = "child2"
        node:
          name = "child3"
          textNode "Hello, world!"

if isMainModule:
  while true:
    doTest()
```

### PR DESCRIPTION
**Pull request by ringabout**
fixes #21171

### CODE REVIEW

File: compiler/liftdestructors.nim [hunk 1]
```diff
@@ -632,6 +637,8 @@ proc atomicRefOp(c: var TLiftCtx; t: PType; body, x, y: PNode) =
       cond = callCodegenProc(c.g, "nimDecRefIsLastCyclicStatic", c.info, tmp, typInfo)
     else:
       cond = callCodegenProc(c.g, "nimDecRefIsLastCyclicDyn", c.info, tmp)
+  elif isInheritableAcyclicRef:
+    cond = callCodegenProc(c.g, "nimDecRefdIsLastDyn", c.info, x)
```
Comment (ringabout) [inline compiler/liftdestructors.nim]:
```suggestion
    cond = callCodegenProc(c.g, "nimDecRefIsLastDyn", c.info, x)
```

File: lib/system/orc.nim [hunk 2]
```diff
@@ -485,6 +485,19 @@ proc nimDecRefIsLastCyclicDyn(p: pointer): bool {.compilerRtl, inl.} =
     #if cell.color == colPurple:
     rememberCycle(result, cell, cast[ptr PNimTypeV2](p)[])
 
+proc nimDecRefIsLastDyn(p: pointer): bool {.compilerRtl, inl.} =
+  if p != nil:
+    var cell = head(p)
+    if (cell.rc and not rcMask) == 0:
+      result = true
+      #cprintf("[DESTROY] %p\n", p)
```
Comment (bung87) [inline lib/system/orc.nim:493]:
```suggestion
```

File: lib/system/orc.nim [hunk 3]
```diff
@@ -485,6 +485,19 @@ proc nimDecRefIsLastCyclicDyn(p: pointer): bool {.compilerRtl, inl.} =
...
+    else:
+      dec cell.rc, rcIncrement
+    #if cell.color == colPurple:
```
Comment (bung87) [inline lib/system/orc.nim:496]:
```suggestion
```

Comment (bung87) [inline lib/system/orc.nim:496]:
```suggestion
```

File: lib/system/orc.nim (hunk 2 above)
Comment (ringabout) [reply lib/system/orc.nim:493]: It is used for debugging these procs and is copied from `nimDecRefIsLastCyclicDyn`.

**Pull request merged**